compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,505 @@
+        emptyVec, emptyVec, transposed, emptyVec, emptyVec, emptyVec);
+
+    rewriter.eraseOp(unpackOp);
+    return success();
+  }
+};
+
+struct LowerPackResult {
+  memref::ExpandShapeOp expandShapeOp;
+  memref::TransposeOp transposeOp;
+  xilinx::air::DmaMemcpyNdOp dmaOp;
+};
+
+struct LowerUnPackResult {
+  memref::TransposeOp transposeOp;
+  xilinx::air::DmaMemcpyNdOp dmaOp;
+};
+
+FailureOr<LowerPackResult> lowerPack(RewriterBase &rewriter,
+                                     IREE::LinalgExt::PackOp packOp) {
+  // 1. Filter out NYI cases.
+  auto packedMemrefType = packOp.getOutputType();
+  if (llvm::any_of(packOp.getStaticInnerTiles(),
+                   [](int64_t size) { return ShapedType::isDynamic(size); })) {
+    return rewriter.notifyMatchFailure(
+        packOp,
+        "non-static shape NYI, needs a more powerful memref.expand_shape op");
+  }
+
+  Location loc = packOp->getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(packOp);
+
+  // 2. Compute the permutation vector to shuffle packed shape into the shape
+  // before any outer or inner permutations have been applied. The permutation
+  // can be obtained from two permutations:
+  //   a) Compute the permutation vector to move the last `numPackedDims` into
+  //      the `innerPosDims` of a shape of rank `packedRank`.
+  //   b) Compute the permutation vector to move outer dims if the pack op
+  //      has outer_dims_perm.
+  // Apply (b) permutation on (a) permutation to get the final permutation.
+  int64_t numPackedDims = packOp.getInnerDimsPos().size();
+  int64_t packedRank = packedMemrefType.getRank();
+  auto lastDims = llvm::to_vector(
+      llvm::seq<int64_t>(packedRank - numPackedDims, packedRank));
+  PackingMetadata packingMetadata = computePackingMetadata(
+      packedMemrefType.getRank(), packOp.getInnerDimsPos());
+  SmallVector<int64_t> innerPositionsPerm = computePermutationVector(
+      packedRank, lastDims, packingMetadata.insertPositions);
+
+  SmallVector<int64_t> outerPos = packingMetadata.outerPositions;
+  ArrayRef<int64_t> outerPerm = packOp.getOuterDimsPerm();
+  if (!outerPerm.empty()) applyPermutationToVector(outerPos, outerPerm);
+  SmallVector<int64_t> outerPositionPerm = computePermutationVector(
+      packedRank, packingMetadata.outerPositions, outerPos);
+
+  SmallVector<int64_t> packedToStripMinedShapePerm = innerPositionsPerm;
+  applyPermutationToVector(packedToStripMinedShapePerm, outerPositionPerm);
+
+  // 3. Compute the stripMinedShape: this is the packed shape before any outer
+  // or inner permutations have been applied.
+  SmallVector<int64_t> stripMinedShape(packedMemrefType.getShape());
+  applyPermutationToVector(stripMinedShape, packedToStripMinedShapePerm);
+
+  // 4. Expand from the padded result to the stripMinedShape.
+  auto reshapeOp = rewriter.create<memref::ExpandShapeOp>(
+      loc, stripMinedShape, packOp.getInput(), packingMetadata.reassociations);
```

Comment:
To be more specific, you can check if outer dims are all ones or not here. If so, you create rank-reduced subview. If not, you create memref::ExpandShapeOp. Then you are able to drop the other pattern.

Reply:
Thanks for the explanation. I did a code refactor today and dropped the `GeneralizePack(UnPack)OpPattern` functions.